### ISSUE TITLE
Bug: incorrect parameters passed only in release build linux-x64

### ISSUE THREAD

**Issue by abelykh0** [dotnet/coreclr#22330]
## On a specific project the release build is passing incorrect parameters under linux-x64. 

Platform: Linux x64. I tested on Debian GNU/Linux 9 (stretch) and Red Hat Enterprise Linux Server release 
7.5 (Maipo).

## Steps to replicate:

0. Install .Net Core 2.2 on linux
1. Use this project: https://github.com/abelykh0/NetCoreIssue
2. Open the solution LinuxNetCoreBug.sln in Visual Studio 2017
3. Right click on Publish project, Publish, then choose "Release linux-x64". Change the folder if needed (it is currently set to C:\Projects\Issue). Click Publish
4. Run dotnet exec LinuxNetCoreBug.dll on linux

Here is the implementation of the abstract method (the implementation is in the `LoadedLibrary` assembly, which is loaded dynamically with `Assembly.LoadFrom` by the main program):

```
protected override IList<Guid> DbFileSystemQuery(
    int pageNumber,
    int pageSize,
    ICollection<Guid> queryTargetId,
    FileSystemQueryOptions queryOptions,
    IList<Tuple<FileSystemQueryField, SortDirection>> orderBy,
    ICollection<FileSystemQueryFilterRule> filter)
{
    Console.WriteLine(orderBy.GetType().ToString());
    Console.WriteLine(filter.GetType().ToString());
    Console.WriteLine(orderBy == filter);
    return null;
}
```

## Actual result (on linux x64):
> CoreLibrary.FileSystemQueryFilterRule[]
> CoreLibrary.FileSystemQueryFilterRule[]
> True

This output shows that the `orderBy` and `filter` parameters are the same object. However, this should be impossible because the types don’t even work out.

## Expected result (works in debug build or in debug or release build in Windows):
> System.Tuple`2[CoreLibrary.FileSystemQueryField,CoreLibrary.SortDirection][]
> CoreLibrary.FileSystemQueryFilterRule[]
> False

**Comment by RussKeldorph** [comment]
@dotnet/jit-contrib Does this ring a bell as something that might have already been fixed?

@abelykh0 Are you requesting a fix to 2.2?

**Comment by abelykh0** [comment]
Yes, please

**Comment by abelykh0** [comment]
The workaround exists (using debug DLL), however we would obviously prefer to be able to ship release build to the customers.

**Comment by 0xd4d** [comment]
> It seems to be fixed in 3.0.0 latest preview.

Did you try disabling tiered compilation? Set env var `COMPlus_TieredCompilation=0` and then run the app.

**Comment by abelykh0** [comment]
COMPlus_TieredCompilation=0 doesn't make any difference, still incorrect.

**Comment by 0xd4d** [comment]
Disabling tiered compilation means that your code gets optimized and you get the same buggy result. This bug is not fixed.

**Comment by abelykh0** [comment]
Sorry, I edited the comment. Actually, enabling tiered compilation fixes it. If it is disabled, the issue occurs. The reason it is working on 3.0.0 preview is that the default changed from 0 to 1. If I set COMPlus_TieredCompilation=0, it also fails in 3.0.0.

**Comment by RussKeldorph** [comment]
Ah, right.  I should have realized the difference is tiering being enabled.  We generally wouldn't recommend enabling tiering to work around this, though you might be able to use [MethodImplOptions.NoOptimization](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.methodimploptions?view=netframework-4.7.2) to achieve a similar local effect and work around it.  At any rate, we should investigate what's going on.

**Comment by AndyAyersMS** [comment]
Possibly related to #19256?

**Comment by AndyAyersMS** [comment]
@briansull please update if you find anything -- I am going to start looking at the possibly related #19256.

**Comment by AndyAyersMS** [comment]
Ok, does not seem to be the exactly same bug as #19256 but is in a very similar area: we are trying to stage args to the stack, but already have args on the stack, and we don't do things in the right order.

Here is a simple self-contained repro:
```C#
using System;
using System.Runtime.CompilerServices;

class X
{
    public static int F(int a, int b, Guid c, int d, int e, int f, int g)
    {
        Guid[] z = new Guid[] { c };
        // Bug is here passing params to G: f gets trashed by g.
	return G(a, b, z, d, e, f, g);
    }
    
    // loop here is just to make this method too big to inline
    // if we set [noinline] to effect this, we won't tail call it either.
    public static int G(int a, int b, Guid[] c, int d, int e, int f, int g)
    {
	int r = 0;
	for (int i = 0; i < 10; i++)
        {
            r += f + g;
        }
	return r / 10;
    }

    // No-opt to stop F from being inlined without marking it noinline
    [MethodImpl(MethodImplOptions.NoOptimization)]
    public static int Main()
    {
        int result = F(0, 1, Guid.Empty, 3, 4, 33, 67);
        Console.WriteLine($"Result={result}");
        return result;
    }
}
```
Here `F` tail calls `G` and inadvertently trashes arg `f` when setting up the tail call. Net result is that `g` is passed twice.
```asm
G_M25278_IG03:
       448B6C2448           mov      r13d, dword ptr [rsp+48H]   // fetch 'g'
       44896C2440           mov      dword ptr [rsp+40H], r13d   // pass as last arg.. oops
       8BFB                 mov      edi, ebx
       418BF6               mov      esi, r14d
       488BD0               mov      rdx, rax
       418BCF               mov      ecx, r15d
       458BC4               mov      r8d, r12d
       448B4C2440           mov      r9d, dword ptr [rsp+40H]    // pass 'f' as next to last arg
       48B8509C6C72937F0000 mov      rax, 0x7F93726C9C50

G_M25278_IG04:
       4883C410             add      rsp, 16
       5B                   pop      rbx
       415C                 pop      r12
       415D                 pop      r13
       415E                 pop      r14
       415F                 pop      r15
       48FFE0               rex.jmp  rax

; Total bytes of code 124, prolog size 36 for method X:F(int,int,struct,int,int,int,int):int
; ============================================================
Result=134

```

**Comment by briansull** [comment]
@AndyAyersMS @CarolEidt -  I haven't yet investigated this issue on Linux

**Comment by jashook** [comment]
@AndyAyersMS are these fast tail calls or slow

**Comment by AndyAyersMS** [comment]
Fast.

**Comment by jashook** [comment]
I wonder if this is fixed by https://github.com/dotnet/coreclr/pull/20643.

**Comment by AndyAyersMS** [comment]
Is the latest iteration of that PR in good shape? Can't tell what the failures were anymore as the CI logs have been purged.

**Comment by jashook** [comment]
Testing is in good shape but there were ASM diffs. Let me see if there is not much work to handle merge conflicts.

**Comment by jashook** [comment]
As part of the refactoring work I fixed a few bugs that Bruce and I found on the linux path which is why I bring it up.

**Comment by jashook** [comment]
Can easily test, looking now.

**Comment by jashook** [comment]
It is unrelated.

**Comment by jashook** [comment]
Strangely noticed that changing the guid to two ints or an eight byte struct will cause this to not repro. The blockop seems to be required.

**Comment by jashook** [comment]
This bug is a dup of https://github.com/dotnet/coreclr/issues/12468. The underlying problem is that nCalleeArgs and nCallerArgs is not necessarily the same for caller/callees with >8 byte structs. Therefore we cannot back the fundamental assumption made here: https://github.com/dotnet/coreclr/blob/master/src/jit/lower.cpp#L2024. This particular example will blow pas the noway assert because nCallerArgs == nCalleeArgs; however, this case should not be fastTailCallable on linux until this assumption that caller and callee args have a 1 to 1 relationship is removed in LowerFastTailCall. The fix is to report false in fgCanFastTailCall if either the caller or callee has >8 byte structs && stack arguments.

**Comment by jashook** [comment]
In this specific case LowerFastTailCall should create a tmp to copy callee arg 6 to caller arg 5

**Comment by jashook** [comment]
Will pr the fix.

**Comment by abelykh0** [comment]
Not sure if it matters, but in the original replication there are no struct args

**Comment by AndyAyersMS** [comment]
There actually is one -- a `Guid` is a struct. The method below in your repro corresponds to `F` above:

https://github.com/abelykh0/NetCoreIssue/blob/64578256b67e24929ba5acb12084e494681b69d5/Library/Class1.cs#L75-L85

**Comment by abelykh0** [comment]
OK, great. Thanks.

**Comment by RobSiklos** [comment]
What are the chances this fix will make it into some future 2.2 update?

**Comment by echesakovMSFT** [comment]
@jashook FYI, GitHub_22330 has failed in https://dev.azure.com/dnceng/public/_build/results?buildId=110518&view=logs in Ubuntu.1804.Amd64.Open-x64
